core/client/fs/src/main/java/alluxio/client/block/stream/GrpcDataWriter.java

```diff
@@ -0,0 +1,398 @@
+import com.google.common.base.Throwables;
+import com.google.protobuf.ByteString;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data writer that streams a full block or a UFS file to a gRPC data server.
+ *
+ * Protocol:
+ * 1. The client streams data chunks (start from pos 0) to the server. The client pauses if the
+ *    client buffer is full, resumes if the buffer is not full.
+ * 2. The server reads chunks from the stream and writes them to the block worker. See the server
+ *    side implementation for details.
+ * 3. The client can either complete or cancel the stream to end the write request. The
+ *    client has to wait for the comeplte or cancel response from the data server to make
+ *    sure that the server has cleaned its states.
+ * 4. To make it simple to handle errors, the stream is closed if any error occurs.
+ *
+ * NOTE: this class is NOT threadsafe. Do not call cancel/close while some other threads are
+ * writing.
+ */
+@NotThreadSafe
+public final class GrpcDataWriter implements PacketWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(GrpcDataWriter.class);
+
+  private static final long WRITE_TIMEOUT_MS =
+      Configuration.getMs(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+  private static final long CLOSE_TIMEOUT_MS =
+      Configuration.getMs(PropertyKey.USER_NETWORK_NETTY_WRITER_CLOSE_TIMEOUT_MS);
+  /** Uses a long flush timeout since flush in S3 streaming upload may take a long time. */
+  private static final long FLUSH_TIMEOUT_MS =
+      Configuration.getMs(PropertyKey.USER_NETWORK_NETTY_WRITER_FLUSH_TIMEOUT);
+
+  private final FileSystemContext mContext;
+  private final BlockWorkerClient mClient;
+  private final WorkerNetAddress mAddress;
+  private final long mLength;
+  private final WriteRequestCommand mPartialRequest;
+  private final ClientCallStreamObserver<WriteRequest> mRequestStream;
+  private final long mPacketSize;
+
+  /**
+   * Uses to guarantee the operation ordering.
+   *
+   * NOTE: {@link StreamObserver} events are async.
+   * gRPC worker threads executes the response {@link StreamObserver} events.
+   */
+  private final ReentrantLock mLock = new ReentrantLock();
+
+  @GuardedBy("mLock")
+  private long mPosWritten;
+
+  /**
+   * The next pos to queue to the buffer.
+   */
+  @GuardedBy("mLock")
+  private long mPosToQueue;
+  @GuardedBy("mLock")
+  private Throwable mError;
+  @GuardedBy("mLock")
+  private boolean mDone;
+  @GuardedBy("mLock")
+  private boolean mEOFSent;
```

Comment:
Is there no way to simplify all this state with grpc? It would be nice to have a simpler way to implement rpcs with the new framework.

Reply:
The write path is a bi-direction streaming API due to the need to send data chunks and receive multiple responses (flush, complete, etc.). A bi-directional streaming API is always asynchronous and therefore requires something to communicate the state between the gRPC event thread and the caller thread. Unfortunately gRPC stream observer does not expose any stream state except for `isReady`, so we have to do our own state tracking.